### ISSUE TITLE
[C++] Improve Substrait consumer example

### ISSUE THREAD

**Issue by ianmcook** [apache/arrow#33956]
### Describe the enhancement requested

The example at [cpp/examples/arrow/engine_substrait_consumption.cc](https://github.com/apache/arrow/blob/master/cpp/examples/arrow/engine_substrait_consumption.cc) has some shortcomings that make it less useful than it could be for demonstrating how Acero can consume and execute Substrait plans: 

- It does not collect the result into a table and display it (which is quite disappointing after one struggles to get the example running)
- The `add` function is included in the extensions but not used
- The extension references and anchors are silly
- Some of the code and comments seems to conflate the terms "consume" (which should apply to a plan) and "sink" (which should apply to data)
- The user needs to find/create a Parquet file to run it
- `#include <arrow/engine/api.h>` needs to be changed to `#include <arrow/engine/substrait/api.h>`

### Component(s)

C++

**Comment by westonpace** [comment]
> The extension references and anchors are silly

That was partly by design to emphasize that these can be any arbitrary integers. I find it makes it much more obvious how they are related.  For example, if we used `0` instead of `7` then it would not be obvious that `extension_uri_reference` is a reference to `extension_uri_anchor`.

> #include <arrow/engine/api.h> needs to be changed to #include <arrow/engine/substrait/api.h>

That seems a little surprising to me.  What error did you get using `<arrow/engine/api.h>?

**Comment by ianmcook** [comment]
>> The extension references and anchors are silly
>
> That was partly by design to emphasize that these can be any arbitrary integers. I find it makes it much more obvious how they are related.  For example, if we used `0` instead of `7` then it would not be obvious that `extension_uri_reference` is a reference to `extension_uri_anchor`.

Ah, got it, fair enough.

>> #include <arrow/engine/api.h> needs to be changed to #include <arrow/engine/substrait/api.h>
>
>That seems a little surprising to me.  What error did you get using `<arrow/engine/api.h>`?

That file does not exist when I build and install the Arrow library (using a recent dev commit) :
```
% find /usr/local/include/arrow/engine 
/usr/local/include/arrow/engine
/usr/local/include/arrow/engine/substrait
/usr/local/include/arrow/engine/substrait/relation.h
/usr/local/include/arrow/engine/substrait/pch.h
/usr/local/include/arrow/engine/substrait/extension_types.h
/usr/local/include/arrow/engine/substrait/type_fwd.h
/usr/local/include/arrow/engine/substrait/options.h
/usr/local/include/arrow/engine/substrait/extension_set.h
/usr/local/include/arrow/engine/substrait/api.h
/usr/local/include/arrow/engine/substrait/test_plan_builder.h
/usr/local/include/arrow/engine/substrait/util.h
/usr/local/include/arrow/engine/substrait/visibility.h
/usr/local/include/arrow/engine/substrait/serde.h
```

**Comment by ianmcook** [comment]
>The user needs to find/create a Parquet file to run it

To solve this, maybe we could just add a comment saying something like this:

To create a Parquet file `example.parquet` to use with this example, run this Python code:
```py
import pyarrow as pa
import pyarrow.parquet as pq
pq.write_table(pa.Table.from_pydict({'i': [1,2,3,4], 'b': [True,False,True,False]}), 'example.parquet')
```

**Comment by vibhatha** [comment]
take

**Comment by westonpace** [comment]
> That file does not exist when I build and install the Arrow library (using a recent dev commit) :

Oops.  https://github.com/apache/arrow/pull/33964

**Comment by vibhatha** [comment]
Looking into the example after many iterations since the initial PR, I think the most intuitive and helpful example could be to use a data source with a `NamedTableProvider`. We already have many test cases wrapped around that. 

This means existing code could completely change. What do you think? 

cc @westonpace @ianmcook

**Comment by ianmcook** [comment]
@vibhatha sounds OK to me, but I defer to Weston

**Comment by westonpace** [comment]
Agreed.  Not only is it simpler (don't need to rely on user to create a parquet file) but it demonstrates even more features :)

**Comment by vibhatha** [comment]
I will modify accordingly. Thanks for the feedback.

**Comment by kabicm** [comment]
Is there any update on this? What should be the final example showing how we can consume substrait plans with Acero in C++?
